diff --git a/src/oic/oic/__init__.py b/src/oic/oic/__init__.py
index 3ac8f14..bd6677c 100644
--- a/src/oic/oic/__init__.py
+++ b/src/oic/oic/__init__.py
@@ -268,9 +268,12 @@ rt2gt = {
 }
 
 
-def response_types_to_grant_types(response_types):
+def response_types_to_grant_types(response_types, kwargs):
     _res = set()
 
+    if 'grant_types' in kwargs:
+        _res.update(set(kwargs['grant_types']))
+
     for response_type in response_types:
         _rt = response_type.split(' ')
         _rt.sort()
@@ -1353,7 +1356,7 @@ class Client(oauth2.Client):
 
         if 'response_types' in req:
             req['grant_types'] = response_types_to_grant_types(
-                req['response_types'])
+                req['response_types'], kwargs)
 
         return req
 
